### PR TITLE
TokenField: Add disabled prop

### PR DESCRIPTION
**Pull request by ebinnion**
Related to #3441. While submitting invites, we want to disable the TokenField so that changes can not be made.

<strike>So far, this PR does disable the TokenField, but there are no styling changes as of yet.

For the main TokenField component, I plan on duplicating the disabled styled for inputs, but I am not sure how to handle the token styles. Perhaps a light gray color? cc @rickybanister </strike>

To test:
- Checkout `update/token-field-disabled` branch
- Go to `/people/new/$site` and submit invite form
- Upon submitting, `TokenField` component at top of invite form should disable while API call is being made

Alternatively, you can open `invite-people/index.jsx` and force the `disabled` prop of `TokenField` to be true.

### CODE REVIEW

**Comment by rickybanister** [comment]
I'm not sure I understand, wouldn't it only be 'disabled' when you have to address a validation problem? In that case we should not make all the existing tokens appear 'disabled' as well. 

Since the actual input in this case is not the bounding box, but an invisible input, we may be able to just hide that element. I'll take the branch for a spin tomorrow and see if that makes sense.

**Comment by ebinnion** [comment]
I haven't completely integrated this branch with the invite validation yet. To test the `disabled` state, you can modify [this line](https://github.com/Automattic/wp-calypso/pull/3572/files#diff-ace89b667e7bd9a57e091cdd09278e63R63) to be true.

Then you can go to `/people/new/$site` or `/devdocs/design/token-fields` and test the component.

It is pretty rough, but I wanted to get something up so that we can discuss what disabling actually meant. I think there are two cases where we will want to disable:
- When saving the invites form, prevent removing or adding tokens
- When there is a validation error, prevent adding more tokens

**Comment by ebinnion** [comment]
Alrighty, I've taken a few more passes at this, and I've got something that is starting to feel OK to me.

We disable the `TokenField` as soon as there is any error **OR** if the form has been submitted. 

When the `TokenField` is disabled, it takes on similar styles to a disabled `textarea`. The only action that can be taken on the `TokenField` is that a user can click the "x" to remove errored tokens. 

Screenshots:

screen shot 2016-03-02 at 4 33 54 pm
screen shot 2016-03-02 at 4 39 02 pm

**Note:** that I added the `borderless` and regular token styles here. For our purposes in people, we use the top `borderless` style. Also note that the errors are a bit wordy. I plan to address those in a separate PR.

File: client/components/token-field/token.jsx [hunk 1]
```diff
@@ -50,14 +53,16 @@ export default React.createClass( {
...
-					className="token-field__remove-token noticon noticon-close-alt"
-					onClick={ this._onClickRemove } />
-			{ tooltip &&
-				<Tooltip context={ this } status={ status } isVisible={ true } position="bottom">
-						{ tooltip }
-				</Tooltip>
-			}
+				{ ! this.props.disabled &&
+					<span
+						className="token-field__remove-token noticon noticon-close-alt"
```
Comment (alternatekev) [inline client/components/token-field/token.jsx]: Why are we using Noticons here? These should be Gridicons.

**Comment by alternatekev** [comment]
I know that comma will trigger the tokenization, but I think a `space` keypress should also trigger it.

**Comment by alternatekev** [comment]
I fixed some SCSS hierarchy and pushed the x icon down 1px when it's used in the borderless state as it was a little too high.

**Comment by ebinnion** [comment]
I just finished rebasing this PR against master and cleaning up some extraneous commits that somehow found their way in. :scream: 

I chatted with @alternatekev in Slack today. He pointed out that removing the "x" icons when disabling the `TokenField` is pretty janky. I agree.

Tomorrow, I'll pick this back up and address:
- Gridicon instead of noticon for "x" in token
- Don't remove "x", just disable it

**Comment by alternatekev** [comment]
I may have screwed up my rebase :(

**Comment by rickybanister** [comment]
It feels good to me. 

I agree with @alternatekev about `space` triggering a token—I expected that. 

I don't agree with him about moving the x down, he created a jump when you enter the first token, the entire field gets 1px taller from moving down the x. We should either absolutely position it to prevent the change in height, or revert. 

Once we address that slight regression we can merge.

**Comment by alternatekev** [comment]
Ah I missed the jump. I can update the CSS to do better positioning.

File: client/components/token-field/token.jsx [hunk 2]
```diff
@@ -52,12 +55,12 @@ export default React.createClass( {
...
-					onClick={ this._onClickRemove } />
-			{ tooltip &&
-				<Tooltip context={ this } status={ status } isVisible={ true } position="bottom">
-						{ tooltip }
-				</Tooltip>
-			}
+					onClick={ ! this.props.disabled && this._onClickRemove } />
+				{ tooltip &&
+					<Tooltip context={ this } status={ status } isVisible={ true } position="bottom">
+							{ tooltip }
```
Comment (lezama) [inline client/components/token-field/token.jsx]: looks like there's an extra `tab` here?

File: client/components/token-field/index.jsx [hunk 3]
```diff
@@ -90,21 +101,30 @@ var TokenField = React.createClass( {
...
 		var classes = classNames( 'token-field', {
-			'is-active': this.state.isActive
+			'is-active': this.state.isActive,
+			'is-disabled': this.props.disabled
 		} );
 
+		var tokenFieldProps = {
+			ref: 'main',
+			className: classes,
+			tabIndex: '-1'
```
Comment (enejb) [inline client/components/token-field/index.jsx]: Why do we need the tabIndex: -1 here?

File: client/components/token-field/token-input.jsx [hunk 4]
```diff
@@ -32,6 +32,7 @@ var TokenInput = React.createClass( {
 				onBlur={ this.props.onBlur }
 				onChange={ this._onChange }
 				className="token-field__input"
+				disabled={ this.props.disabled }
```
Comment (enejb) [inline client/components/token-field/token-input.jsx]: We should add the props to the default props and propTypes.

File: client/components/token-field/index.jsx (hunk 3 above)
Comment (ebinnion) [inline client/components/token-field/index.jsx]: I'm not 100% sure. But, the prop was there before. I just changed how it is added.

File: client/my-sites/people/invite-people/index.jsx [hunk 5]
```diff
@@ -89,8 +93,18 @@ const InvitePeople = React.createClass( {
 			return value;
 		} );
 
+		const filteredErrors = pickBy( errors, ( error, key ) => {
```
Comment (enejb) [inline client/my-sites/people/invite-people/index.jsx]: Why is errors an object but success an array?

Comment (ebinnion) [inline client/my-sites/people/invite-people/index.jsx]: Because, with errors, we get the value as well as an explanation as to why the validation failed. With successes, we don't need the explanation.

File: client/components/token-field/index.jsx (hunk 3 above)
Comment (enejb) [inline client/components/token-field/index.jsx]: I think divs don't support tabIndexs. I think we can just remove it.

File: client/my-sites/people/invite-people/index.jsx (hunk 5 above)
Comment (enejb) [inline client/my-sites/people/invite-people/index.jsx]: Could we not deal with an array of errors still? [ { }, { } ].
Then I think things would be simpler. Sorry I just don't quite understand.

**Comment by enejb** [comment]
Beside the ability to use a space as well as comma and a tab. I think this works quite well.

File: client/my-sites/people/invite-people/index.jsx (hunk 5 above)
Comment (ebinnion) [inline client/my-sites/people/invite-people/index.jsx]: We could. Which would make it a bit easier to use `filter` for both errors and successes in the case above.

But, it would also slightly complicate the error logic in `getTokensWithStatus()`, which is currently: `if ( errors && errors[ value ] )`.

That would probably become something like `if ( some( errors, ( error ) => { error.value === value } ) )`.

Which is still readable, but I'm not sure that it's worth the effort to change.

File: client/components/token-field/index.jsx (hunk 3 above)
Comment (ebinnion) [inline client/components/token-field/index.jsx]: When I search, I'm seeing articles that suggest we should leave `tabIndex: -1` here.

`If given a value of "-1", the element can't be tabbed to but focus can be given to the element programmatically`

**Comment by enejb** [comment]
The PR is works well for me. :+1: :shipit:

**Comment by ebinnion** [comment]
Thanks for the review. I created #3864 to tokenize on space.
